### PR TITLE
Block variations transformations

### LINKED ISSUE #25231: Add a way to transform between block variations
The navigation block has two variations: horizontal and vertical. These have to be variations because styles are CSS only and we need to alter one of the block's attributes to change how the movers behave depending on orientation.

However this is not very intuitive and it's not easy to understand the difference and the "why", and for many instances its has been suggested to "merge the two variations" (e.g. #23226). As there is no way to "merge" the variations, the proxy would be something which behaves like styles and appears to users as a "setting" of the block.

Therefore it could be so that when a block has multiple variations registered it displays the list of variations in one of the two places occupied currently by styles:

Screenshot 2020-09-11 at 10 38 43

The new API should be able to partially apply new block attributes, preserving content.

In the case of the navigation example, starting from:
```
<!-- wp:navigation {"orientation":"vertical"} -->
<!-- wp:navigation-link {"label":"Ut atque aut nobis","type":"custom","id":1976} -->
<!-- wp:navigation-link {"label":"Sample Page","type":"custom","id":1979} /-->
<!-- /wp:navigation -->
```
... after applying the operation the result is:

```
<!-- wp:navigation {"orientation":"horizontal"} -->
<!-- wp:navigation-link {"label":"Ut atque aut nobis","type":"custom","id":1976} -->
<!-- wp:navigation-link {"label":"Sample Page","type":"custom","id":1979} /-->
<!-- /wp:navigation -->
``` 

This new feature is also useful for block based themes as it allows themes to apply their own attribute customizations to existing data saved in template parts.

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Resolves: https://github.com/WordPress/gutenberg/issues/25231

This PR is the first step for handling block variation transformations. It introduces a `transform` new option for `scope` field in block variations. This way we can control how and when to handle these transformations, as this behavior doesn't always make sense for all variations for every block.

In this PR we support block variations that change only attributes, like the `Navigation` block (which is the only one adjusted for testing reasons). The design is not finalized yet and I've implemented it with a simple select control for now, just to test how this feels.

The second step in a follow up PR will be to handle transformations with `InnerBlocks` like Query or Columns that will need an API to manually handle such block variation transformations. For example what happens when we have Columns (4) and we want to expose the transform to Columns (3) etc... 

## Testing Instructions

1. Insert and init a Navigation Block (create from top level pages)
2. Select the Navigation block (not any child of it)
3. The select box is shown below the Block Card after selecting the block based on the position that seems to be prevailing at issue's discussions(https://github.com/WordPress/gutenberg/issues/25231#issuecomment-720864900)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: docs/designers-developers/developers/block-api/block-registration.md
Comment (talldan) [inline docs/designers-developers/developers/block-api/block-registration.md:267]: Would be great to see further definition of these options, I understand `inserter` and `transform` would control whether the block is shown in the inserter or as a transform, but wasn't straight away sure what the `block` option does. 

Is that for something like the columns block where we see the variations in a placeholder?

Comment (ntsekouras) [reply docs/designers-developers/developers/block-api/block-registration.md:267]: I have this in mind yes and will update docs where needed later. For the record:
- inserter - shows variation in inserter
- block - used by Placeholders like you mentioned in Columns
- transform - will be shown in the component for block variation transformations

I do also feel that this abstraction here https://github.com/WordPress/gutenberg/blob/b34ecc23f89066e3e7bb710136efa46e29ebe025/packages/blocks/src/store/selectors.js#L87 makes more than it should be and this surfaced with the addition of `transform` as `scope`.

>When not provided, it assumes all available scopes.

With `transform` we want to have this scope and doesn't fall in the rest cases that `no scope means every scope`.  What do you think about changing this to have an option like `strictFiltering` which will default to `false` for back compat?

Comment (gziolo) [reply docs/designers-developers/developers/block-api/block-registration.md:267]: It is about time to convert this list into a proper JSDoc type and list all options according to the comment you shared. By the way, should it be plural `transforms` as in the target place where it's going to be used?

> > When not provided, it assumes all available scopes.
> With `transform` we want to have this scope and doesn't fall in the rest cases that no scope means every scope. What do you think about changing this to have an option like `strictFiltering` which will default to `false` for back compat?

Do we really need a new option there? Wouldn't it be enough to set the list of scopes to `[ 'block', `inserter` ]` if the scope param is not provided?

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/index.js:22]: It definitely should be handled inside the `getBlockVariations` implementation.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/index.js:35]: Is it possible that a variation wouldn't have a title? I thought it is mandatory.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/index.js:40]: How is it possible? Shouldn't be the option that acts like a label disabled so users couldn't pick them?

File: docs/designers-developers/developers/block-api/block-registration.md
Comment (gziolo) [inline docs/designers-developers/developers/block-api/block-registration.md:267]: Now I see that `WPBlockVariationScope` is a thing so maybe we should update to.
```suggestion
- `scope` (optional, type `WPBlockVariationScope[]`) - the list of scopes where the variation is applicable. When not provided, it assumes all available scopes. Available options: `block`, `inserter`, `transform`.
```

**Review by gziolo — COMMENTED** [review]
It's going in the right direction. It feels like it should be fine as an initial proposal to get it tested in the plugin.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js:35]: I have to check that again but I saw in `Search` variation that it doesn't have one.. https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/search/variations.js

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js:22]: This is here because of my comment for the `getBlockVariations`. 

>Do we really need a new option there? Wouldn't it be enough to set the list of scopes to [ 'block', 'inserter' ] if the scope param is not provided?

Where? In `getBlockVariations`? If yes it still feels that the abstraction will not be good to me to have specific business logic per `scope` passed. If in the future we add another `scope` that needs to be filtered exclusively we should add logic for this as well there?

I think this:
```
return variations.filter( ( variation ) => {
		return ! variation.scope || variation.scope.includes( scope );
	} );
```
was there for convenience because `block` as a scope wasn't affecting anything else.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js:40]: This is just temporary as the design will change and probably will not be a simple select.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/index.js:22]: How about this:
```diff
diff --git a/packages/blocks/src/store/selectors.js b/packages/blocks/src/store/selectors.js
index 21e88311b5..34bd862e89 100644
--- a/packages/blocks/src/store/selectors.js
+++ b/packages/blocks/src/store/selectors.js
@@ -90,7 +90,11 @@ export function getBlockVariations( state, blockName, scope ) {
 		return variations;
 	}
 	return variations.filter( ( variation ) => {
-		return ! variation.scope || variation.scope.includes( scope );
+		return (
+			( ! variation.scope &&
+				[ 'block', 'inserter' ].includes( scope ) ) ||
+			variation.scope.includes( scope )
+		);
 	} );
 }
```

Would it work?

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/index.js:35]: I'm sure that the variation for search inherits the title from the block type then.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js:22]: Yes a concept like that would work (I didn't test but though something similar to implement at first), but 
>it still feels that the abstraction will not be good to me to have specific business logic per scope passed. If in the future we add another scope that needs to be filtered exclusively we should add logic for this as well there?

Do you think adding these kind of checks kind of making the abstraction feel wrong?

Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/index.js:22]: In what way wrong? It only ensures backward compatibility, otherwise, all newly added scopes will be only matched when explicitly listed in the list passed in `scope` field. Do I miss anything here? An example where it won't work as intended would be very helpful.

Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js:22]: I never said it wouldn't work. It will. I just said that the abstraction didn't seem right to me now that we have more scopes (this `! variation.scope`). Backward compatibility is a thing though... I will make that change as you suggested.

Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/index.js:22]: I simply didn't take into account that defaulting to all scopes when no scopes are set would cause troubles. I think it all started with `block` as the only supported scope, then we added `inserter`.

**Comment by paaljoachim** [comment]
This is what I see:

Navigation-Variations

An overview:
Variation drop down is seen in the sidebar when the Navigation startup screen is present.
Variation drop down is also seen when menu links are in place.
This gives the user the option to switch between horizontal and vertical mode, before or after having added the menu items. 


I have a few comments that is outside the scope of this PR, but they show up as a natural extension of adding the Variation drop down.

We see text for selecting horizontal or vertical. Which makes it pretty clear what it does. At a later time one will likely think of other Navigation variations that can be added to the drop down. Variations where it would be beneficial to use text and icons. I am wondering if we should in general use text and icons for all variations. 

Screen Shot 2020-11-06 at 13 10 41


--

`/` na the Navigation (horizontal) and the Navigation (vertical) shows up. After we get this PR merged it would be natural to just have one general Navigation block that starts out horizontal. The user can select vertical/horizontal in the sidebar drop down.
Screen Shot 2020-11-06 at 13 00 09

**Comment by ntsekouras** [comment]
> Variation drop down is seen in the sidebar when the Navigation startup screen is present.
>/ na the Navigation (horizontal) and the Navigation (vertical) shows up. After we get this PR merged it would be natural to just have one general Navigation block that starts out horizontal. The user can select vertical/horizontal in the sidebar drop down.

Maybe it would be better just to hide it during initialisation.. I'll have to check more how to make it work with every block best, as most of the blocks do not have a Placeholder for creation. I think it'll become more obvious after checking what core blocks will use that feature. 

I'll do that after we have some design in place for the available options.

>am wondering if we should in general use text and icons for all variations.

I agree with that. Icons will be pretty useful.

**Comment by draganescu** [comment]
I wonder, probably silly, why do variations need to opt in to be available as transforms? Can't we default to having all variations available?

**Comment by ntsekouras** [comment]
Hey @draganescu! 

>I wonder, probably silly, why do variations need to opt in to be available as transforms? Can't we default to having all variations available?

There are more nuances to this, especially when having variations with InnerBlocks just like Columns and Query that we have to handle the content as well and probably have to do it manually with some kind of `block variation transform function`(second step in a follow up PR).

For example when we have a `3 columns Columns Block` should we show the transform to `2 Columns`? It needs custom handling for the content. Another example is the `embed` block which tries to auto transform in block variations, so it wouldn't make sense at all to show this list.

File: packages/blocks/src/store/test/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/test/selectors.js:293]: Nit: You can inline `expectedVariationNames` to make it easier to read.

File: packages/blocks/src/store/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/selectors.js:96]: I don't like this implementation proposed because it depends on the value provided.

Based on the current API, it should be fine to assume that when the `scope` is set, and the variation has no `scope` set then the value used for operations should be `[ 'block', 'inserter' ]`.

```js
return variations.filter( ( variation ) => {
 	// For backward compatibility reasons, variation's scope defaults to `block` and `inserter` when not set.
 	return ( variation.scope || [ 'block', 'inserter' ] ).includes( scope );
 } );
```
It might be enough to code it like the above.

File: packages/blocks/src/store/test/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/test/selectors.js:273]: This comment is no longer valid, it can be actually removed.

File: packages/blocks/src/store/test/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/test/selectors.js:258]: Is this test still necessary?

**Review by gziolo — APPROVED** [review]
Code wise, this PR is ready to go. There are two things that need to happen to consider it done:
- documentation for `BlockVariationTransforms` component needs to be polished
- design review is required for the first iteration

I will leave the final call to other folks but from myself it's ✅

File: packages/block-editor/src/components/block-variation-transforms/README.md
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/README.md:42]: This example made me think, it feels like `blockName` is redundant in the public API of `BlockVariationTransforms`. You can compute it based on `selectedBlockClientId`. It would simplify this example as you would only need to call `getSelectedBlockClientId` inside `useSelect` call.

Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/README.md:42]: You're right. I'll update the code.

**Comment by paaljoachim** [comment]
I believe it will benefit the user to use thumbnails for all variations. As one would get text and a visual image. It would keep a consistency in relation to the UI for variations that all use thumbnails. As I mention in this comment: https://github.com/WordPress/gutenberg/issues/25231#issuecomment-721254467

In relation to the Navigation variation drop down in this PR.

Without thumbnails:
Nav-block-Variations-no-thumbnails

With thumbnails:
Nav-block-Variation-with-thumbnails

In relation to the horizontal and vertical variation having a thumbnail might not be that important, but if one begins to add other variations to the Navigation block then thumbnails might be needed to give a visual impression. Picking and choosing when to use thumbnails and not would break the consistency. Which means even though some variations are easy to understand with the text it would be better to have all variations using a thumbnail approach. Somewhat similar to the Styles Panel. Some people understand better with visual and some better with text, using both can be very helpful.
(Btw most of this is a repeat of what I mention in the comments I made in the issue: https://github.com/WordPress/gutenberg/issues/25231)

@ItsJonQ @jasmussen

**Comment by jasmussen** [comment]
Question: why does the navigation block appear as two blocks, requiring an entirely new UI to transform between them, when we already have this?

Screenshot 2020-11-10 at 08 14 12

This seems like a lot of added complexity, for reasons I can't really spot.

**Comment by jasmussen** [comment]
Looks like my question has already been discussed at length, and answered in https://github.com/WordPress/gutenberg/issues/25231#issuecomment-707909261. 

In that light, let's focus the design in on Shaun's mockup at least for now:



Thumbnails are hard, because they scale very badly, having to be created manually. Actual previews, like those in the transform dialog, are an option that can be explored in a future iteration.

**Comment by jasmussen** [comment]
This is what I see in the latest iteration:

transforms

Seems like there's an opportunity in the future to revisit all transforms ([Shaun has lots of good thoughts](https://github.com/WordPress/gutenberg/issues/25231#issuecomment-691146491)). But in the mean time, and in the spirit of iterating based on based on [this feedback](https://github.com/WordPress/gutenberg/issues/25231#issuecomment-720864900), there are a couple of things:

The dropdown should show the selected variation when collapsed. I.e. instead of saying "Transform to variation", it should say "Horizontal" or "Vertical". 

There's something up with the padding, it's too tight:

Screenshot 2020-11-12 at 11 36 13

It's this CSS that overrides it:

Screenshot 2020-11-12 at 11 36 09

Here it is with its intrinsic padding:

Screenshot 2020-11-12 at 11 36 58

That chevron on the right still looks funky and scaled down, but that looks like a separate issue with the dropdown component itself.

There doesn't seem to be a default choice. I.e. when I insert a horizontal navigation, I'd expect the checkmark to be present for the horizontal choice. 

The menu jumps a bit in width, which is jarring. You probably need to provide a min-width to the popover. 

Finally, instead of saying "Navigation (horizontal)", should it just say "Horizontal"?

**Comment by jasmussen** [comment]
Please note in the above GIF, the actual variation doesn't appear to change in the content. That's the fault of my theme. It works fine in TwentyTwentyOne

**Comment by ntsekouras** [comment]
Thanks for reviewing @jasmussen!

>There doesn't seem to be a default choice. I.e. when I insert a horizontal navigation, I'd expect the checkmark to be present for the horizontal choice.

This is a problem quite difficult/expensive to solve and with no big value, especially when this will include transformations with  `InnerBlocks`.

>The dropdown should show the selected variation when collapsed. I.e. instead of saying "Transform to variation", it should say "Horizontal" or "Vertical".

This change would make the Component difficult to understand what is showing, no? We have no other label. I do believe though that's fine because in combination with the above comment of not knowing the variation exactly (it will be quite obvious with the InnerBlocks variations) it could be not valid information as well. 

>Finally, instead of saying "Navigation (horizontal)", should it just say "Horizontal"?

This is the block variation's description...

**Comment by jasmussen** [comment]
> This is a problem quite difficult/expensive to solve and with no big value, especially when this will include transformations with InnerBlocks.

Is there a hack? Showing whichever variation of block you chose, without it actually being "set"?

For example with Social Links, the default size is "Medium". It hasn't actually been _set_, and you can choose to explicitly set it, which adds the classname. But we still show a checkmark next to medium, because that's the size you see.

> This change would make the Component difficult to understand what is showing, no? We have no other label. I do believe though that's fine because in combination with the above comment of not knowing the variation exactly (it will be quite obvious with the InnerBlocks variations) it could be not valid information as well.

I'm just thinking of this in terms of being a dropdown menu — a dropdown menu should show the selected choice. Anything else is a bit of a departure from the design, and if we can't make it work as such, I'm not sure it's as effective.

File: packages/block-editor/src/components/block-variation-transforms/README.md
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/README.md:38]: I know this API is still experimental, but technically speaking, it doesn't have to be a block that is currently selected so maybe we should update this prop to:
```suggestion
			blockClientId={ selectedBlockClientId }
```

What do you think?

**Comment by ntsekouras** [comment]
>For example with Social Links, the default size is "Medium". It hasn't actually been set, and you can choose to explicitly set it, which adds the classname. But we still show a checkmark next to medium, because that's the size you see.

Yeah, that is handled with default css and can't work in a generic component like this.

>I'm just thinking of this in terms of being a dropdown menu — a dropdown menu should show the selected choice. Anything else is a bit of a departure from the design, and if we can't make it work as such, I'm not sure it's as effective.

I'm not sure what other alternatives we have here. I think this component is more like an `actions` menu. `Transform` to this and that's it. I wasn't really sure about the `check` icon either for that reason. The `check` icon is something more of this what you selected last but it could potentially not be the actual representation..

If we remove the `checked` icon would it make more sense as a design?

File: packages/block-editor/src/components/block-variation-transforms/README.md
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/README.md:38]: It makes sense, yes.

**Comment by gziolo** [comment]
> If we remove the `checked` icon would it make more sense as a design?

I second that idea. The issue with the current proposal is that it will work as intended only when a user triggers the transformation in the dropdown menu. As soon the user navigates away from the visual editor, by opening a code editor mode or going to another page, there is no simple way to detect which transformation was applied. Well, in this particular case, it would be easy to detect, because there is only one attribute that gets updated. However, as @ntsekouras already mentioned, this can quickly become complex to verify when there are inner blocks involved. Regardless, we would need to extend block variations API with a logic that detects the currently applied transform if any. I can imagine a case when variations don't cover all combinations of values that attribute to be modified can cover. 

Aside: rather than using block variations for the Navigation block, we could also add the same dropdown that modified attributes directly ... 😃

**Comment by jasmussen** [comment]
Part of the problem is that this is yet another way to transform blocks. I understand it's meant to solve [this problem](https://github.com/WordPress/gutenberg/issues/25231#issuecomment-707909261):

> The problem that needs solving here is having a way to 'update' a block to another block variation, after its creation. This has nothing to do with where the variations were available in the first place (inserter/placeholder pattern etc..).

In light of that, it seemed like like [this dropdown design](https://user-images.githubusercontent.com/191598/92940791-8a50c700-f41d-11ea-8f24-4b56c8e72735.png) was a decent path forward — relatively minimal but provided a path forward. But that design is inherently a dropdown, so if it doesn't show the current variation in its resting state, the design falls apart a bit.

Pinging a few for some more thoughts: @shaunandrews @kellychoffman

**Comment by ntsekouras** [comment]
> But that design is inherently a dropdown, so if it doesn't show the current variation in its resting state, the design falls apart a bit.

Since we can revisit later when more complex transforms will be handled, another alternative is to check the selected variation, if there is a match, by checking the attributes for now. That means that this will be the same as now but will have the matched variation with the `check` icon, when clicking the Dropdown.

I have to note again though that this matching might not be possible/valid with complex variations later.

**Comment by gziolo** [comment]
I had also this idea that I shared with @ntsekouras in a private message. If we assume that those transforms never use the inner blocks template but rather call a callback to apply changes, we could easily detect whether any block variation matches current attributes. There might be a limitation where only one variation could match and it would need at least one attribute provided. In fact, the case we have fits perfectly fine with the description provided. It might not work well in another case, but we better battle test it first rather than overthink.

File: packages/block-editor/src/components/block-variation-transforms/style.scss
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/style.scss:1]: It looks like a lot of hassle to style the toggle button to look like something completely different.

Do we have other components that are customized to look and behave like a select control that allows custom HTML code for items? The one that allows picking a different font size might be a better fit in terms of styling and accessibility support.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/index.js:50]: Since we have the helper, we can return the name or `null` when not found:
```suggestion
		setSelectedValue( getMatchingVariationName(
 			blockAttributes,
 			variations
 		) );
```

I updated the name in this example to `getMatchingVariationName`. Is it possible to set `null` as a value to make it explicit that no match was found?

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-variation-transforms/index.js:19]: It looks quite complex so having a unit test covering it would help optimize the internal implementation.

`find` is one way of approaching it, although it returns the first matching element so in the very unlikely case  where two variations set the same attributes it would always pick the first one. I don't know if we should really care about it, but I included an alternative that would use filter:
```js
if ( ! variations?.length ) return; // or it could be removed by defaulting to an empty array in func param
const matches = variations.filter(
    ( { attributes = {} } ) => {
       const keys =  Object.keys(  attributes );
       return keys.length > 0 && isEqual( attributes, pick( blockAttributes, keys ) );
    }
);
if ( matches.length !== 1 ) return null;
return matches[0];
```

File: packages/block-editor/src/components/block-variation-transforms/style.scss
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/style.scss:1]: I have checked that (`CustomSelectControl`) but would want many more additions to support better handling of content - it's not so generic for now.. I think for now it's okay and we can revisit whether to enhance the CustomSelect or create something else.

Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/style.scss:1]: `CustomSelectControl` is a tiny wrapper over [`downshift.js`](https://github.com/downshift-js/downshift) which isn't opinionated about HTML used:
Screen Shot 2020-11-13 at 13 54 16

I would be very surprised if you wouldn't be able to use it here, but I agree that you might need to apply some tweaks as well. The benefit is that it better reflects its purpose from the accessibility point of view.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js]: >Is it possible to set null as a value to make it explicit that no match was found?

Why `null`. Don't you think `undefined` is okay as well? check: https://github.com/WordPress/gutenberg/pull/26687/commits/853b184d50e2e069db74ad26b4450fc36c69d48c

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js]: >I don't know if we should really care about it

I think that we shouldn't for now.. :) But if you think it's better I can change it to `filter`. Although don't you think `isMatch` of lodash would be good to be used in `filter`? `isEqual` and `pick` implementation do something differently? 

I'm writing unit tests now.

File: packages/block-editor/src/components/block-variation-transforms/style.scss
Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/style.scss:1]: I started with that, but saw that needed enough tweaks and seemed rather specific contrary to its name.
1. You think I should change it now or in a follow up PR?
2. The `DropdownMenu` suffers from accessibility point of view? Should we look how to improve this as well? (I haven't dig into accessibility yet)

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/index.js]: Whatever you prefer.

File: packages/block-editor/src/components/block-variation-transforms/index.js
Comment (gziolo) [reply packages/block-editor/src/components/block-variation-transforms/index.js]: It isn't important whether you use `isMatch` or `isEqual`, I wrote this code to present the idea.

Comment (ntsekouras) [reply packages/block-editor/src/components/block-variation-transforms/index.js]: Actually during my tests I realized that `filter` it's better than `find`, even if this is a simple matching function. Thanks for the input @gziolo ! 👍

**Review by gziolo — APPROVED** [review]
Quite a journey 😄 Excellent job tackling this issue, let's see how it works in action 👍

**Pull request merged**
